Source/NSPasteboardItem.m

```diff
@@ -0,0 +1,179 @@
+/** <title>NSPasteboardItem</title>
+
+   <abstract>class for wrapping pasteboard content</abstract>
+
+   Copyright <copy>(C) 2017 Free Software Foundation, Inc.</copy>
+
+   Author: Daniel Ferreira <dtf@stanford.edu>
+   Date: July 2017
+
+   This file is part of the GNUstep GUI Library.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Library General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+
+   You should have received a copy of the GNU Library General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, write to the Free Software Foundation,
+   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/*
+ * FIXME: this class is useless until NSPasteboard implements
+ * NSPasteboardReading + NSPasteboardWriting.
+ */
+
+#import <AppKit/NSPasteboardItem.h>
+#import <Foundation/Foundation.h>
+
+@implementation NSPasteboardItem
+- (id)init
+{
+  if ((self = [super init]))
+    {
+      _providerMap = [[NSMutableDictionary alloc] init];
+      _dataMap = [[NSMutableDictionary alloc] init];
+      _types = [[NSMutableArray alloc] init];
+    }
+
+  return self;
+}
+
+- (NSArray *)types
+{
+  return RETAIN([NSArray arrayWithArray: _types]); 
```

Comment:
For future reference: if you really want to return a copied-but-immutable array with refcount 1, using `[[NSArray alloc] initWithArray: _types]` would avoid adding the object to the autorelease pool in the first place.

Also, you could use `[_types copy]`.

That said, while I am going to approve this, this seems like a leak. If I were calling `[itemFromPasteboard types]`, [based on the method name](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/mmRules.html) I would not expect getting an object I have to `-release` myself. 

I will submit a separate commit removing `RETAIN`.